### PR TITLE
Document and propose some release process stuff

### PR DESCRIPTION
**Pull request by kripken**
This adds docs for the minor version updates, nothing new there but it was never written up, and also a proposal for when to merge to master / a major version update, which needs to be thought about carefully, the text there is just a starting point.

### CODE REVIEW

**Comment by jgravelle-google** [comment]
General idea looks good to me, none of the specifics seem controversial. Having explicit, documented process should be good to iron down *before* doing a major update.

File: docs/process.Markdown [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+Requirements:
+
+ * We should check that there is no obvious breakage on the github CI at least before.
+
+How:
+
+ * The emscripten, emscripten-fastcomp, and emscripten-fastcomp-clang repos should each be updated: the emscripten-version.txt file in each, and a git tag (with the simple version number).
+ * A tag should also be done in the binaryen repo.
+
+Major version update (1.X.Y to 1.X+1.0)
```
Comment (curiousdannii) [inline docs/process.Markdown]: Calling a second level number update a "major" version update is sure to confuse some people...

**Comment by curiousdannii** [comment]
I've come to appreciate semver. But even if you never commit to full semver compliance, having something about API consistency would be good IMO. The recent change to default exports and the change to WASM are both things I think should be "major" updates. If you do want to ever publish emscripten through npm (via wasm etc) then I'd say semver would be a must. Or at least something akin to semver, perhaps bumping the third level for non-breaking updates and second level for breaking updates so that users can use version specifiers as they expect.

>A time of very strong assurance of stability.

What does this actually mean? The docs basically tell everyone to install the latest version through emsdk regardless of what its number is or how stable it might be. In the event that just published "major update" was actually particularly unreliable, would you fix it by bumping the third level or second level number?

**Comment by juj** [comment]
We may want to add a new feature to `emsdk` where `emsdk install latest` is instead replaced by/augmented by a `emsdk install stable`, or something similar - currently installing `latest` always gets the latest precompiled tag, independent of whether we have verified that tag to be stable. And we need to tag to invalidate cache and to rebuild LLVM, so occassionally we need to tag when we don't know that things are necessarily stable.

I'm thinking that Emscripten 2.0 would be something like "targets wasm by default" or something like that functionally substantial.

> In the event that a just published "major update" was actually particularly unreliable, would you fix it by bumping the third level or second level number?

In the past we have done such unreliability fixes by bumping the third level number. Generally second level number has been bumped when we have manually verified all CI tests to be green and have merged to master.

**Comment by dschuff** [comment]
I do sort of like the general idea of semver; we would probably want to think about all the things that might be considered our "API" for that purpose. For example:
* ABI changes like https://github.com/kripken/emscripten-fastcomp-clang/pull/21 (yes)
* Default values for flags (yes)
* New flags (probably not, if they are off by default)
* Bug fixes (no)
etc.

File: docs/process.Markdown (hunk 1 above)
Comment (kripken) [reply docs/process.Markdown]: Fair point. What's a better term? "intermediate" maybe?

**Comment by kripken** [comment]
I also generally like semver. My main concern with changing to a strict version of it is our current testing limitations: we haven't been able to even do a 1.38.0 version and merge that to master for over a year now, just because the bots are never green enough and things are always landing. (The `interactive` tests are a particular problem, I filed #6182.)

If we had reliable automatic tests that allowed us to be sure of stability at a reasonable frequency, I think we should switch to semver. As it is, right now we sort of gauge stability in a relative sense - do all tests pass here on github CI, do they also pass locally, do they also pass on the other 2 places.

**Comment by kripken** [comment]
To make gradual progress here (semver etc. is a longer-term issue), I think the urgent thing to figure out is the process for merging to master / incrementing to 1.38.0. We have some large changes we'd like to do (ABI changes from @sunfishcode, change to wasm by default, etc.) that are blocked on that.

If the text in this PR looks ok, then what I think is left to figure out before landing this is
 * @juj, do you think it's realistic to get your bots green enough? In what is proposed here, we'd open an issue for a merge to master, and we'd need to get all the bots green quickly to do the merge. Right now I see a lot of red there.
 * We need a clear approach for the wasm waterfall, a simple documented way to look at it and see if it's green enough.
 * What to do with the interactive tests, #6182.

**Comment by juj** [comment]
> @juj, do you think it's realistic to get your bots green enough? In what is proposed here, we'd open an issue for a merge to master, and we'd need to get all the bots green quickly to do the merge. Right now I see a lot of red there.

Looking at the errors, these don't seem to be bot failures, but real test failures. We are leaking temporary files related to Closure compiler which fails `test_no_nuthin`, and looks like we've regressed something in SIMD. I can certainly look into these in more detail.

**Comment by kripken** [comment]
I've been working to fix the leak errors on the emscripten buildbot, #6285. That can hopefully resolve the main errors there, but we do still need to figure out a clear answer for when is "good enough" there. I'll keep looking into that.

We also still don't have an answer for the wasm waterfall, as mentioned 2 comments back. Any thoughts?

**Comment by jgravelle-google** [comment]
The wasm waterfall is a bit odd because it's somewhat necessarily flaky due to unrelated upstream LLVM changes that break things, sometimes transiently sometimes not. I'm not sure whether it's a good idea to tie the Emscripten release process to the status of bleeding-edge LLVM, especially when we have no real guarantees of stability for the WebAssembly backend yet (still an Experimental target).

Having said that I'm not sure whether it's a good idea to ignore the signal that it *can* provide, especially with respect to the s2wasm path. The most relevant bits of signal are the "LLVM Torture" tests (actually the [gcc execute torture tests](https://github.com/gcc-mirror/gcc/tree/master/gcc/testsuite/gcc.c-torture/execute)) with the emwasm (wasm backend / s2wasm path) and asm2wasm targets, and the emscripten testsuite tests (mostly the emwasm one, asm2wasm should be covered by some other bots?).

How to tell if it's green enough: no easy answer here. If it's green, it's green enough, but if it's red, how long has it been red, which parts are red, are other LLVM targets red for similar reasons. I have no hard and fast rules here, but I haven't been using its redness as a release blocker, instead looking at it as "fix this".

Is my somewhat scattershot collection of thoughts regarding how the wasm waterfall might relate to Emscripten releases.

**Comment by kripken** [comment]
I see. Ok, then perhaps for now we shouldn't tie emscripten releases to the wasm waterfall. Even more so since the wasm backend isn't the default anyhow - but when we switch to it, at that point we should probably have stable testing for it.

**Comment by dschuff** [comment]
Until recently the wasm waterfall ran asm2wasm tests that the other bots didn't (e.g. the full emscripten test suite with binaryen2). But I think that's no longer the case? So because the wasm backend is not official the waterfall need not block releases (although I would say that when we do want to do a release someone should probably at least check the status of tests for stuff that would be in the release).

Going forward, we will probably want a way to "pin" a revision of the LLVM tools (i.e. specify one that's known to be good). Previously fastcomp was always just pinned by virtue of not being updated but we need a new process to manage changes from upstream LLVM. I guess the current mechanism is to just use matching tags in the LLVM and emscripten repos, but that won't work for upstream LLVM. Whatever mechanism we use has to be really super easy/lightweight (preferably completely automatic) to update and we have to be super disciplined about keeping it up to date, or else the longer it goes, the harder it is.
Maybe a file in the emscripten repo could specify an LLVM revision, but I'm not sure we want to have such frequent commits. It might make sense just to have some sort of specifier in emsdk and the waterfall (and/or whatever other process produces revisions that people actually install), e.g. like a git submodule or a Chrome DEPS file or the json monstronsities that we have in NaCl.

We could split the waterfall or whatever other automated builder along its functions of integration testing (which we still want) vs release building.

But anyway it would be nice to have the following properties:
1) emscripten incoming is as close to the tip of LLVM tree as possible (e.g. some automated LKGR updating)
2) emscripten master doesn't have to be pinned to a release branch of LLVM (which means we need to pin it to a particular intermediate commit; presumably whatever revision was pinned to the one from incoming).
3) ideally the build/release process would have nice github integration (waterfall doesn't have that, don't know about emsdk)
4) other ideas?

**Comment by eclecticdave** [comment]
@dschuff -

* New flags (probably not, if they are off by default)

I realise you made this comment a month ago, but having just added a new flag in my PR it's just occurred to me this isn't so simple.  If you add a new feature behind a flag you don't want to check it into incoming with the flag off - if you do the bots wouldn't run against the new feature.

I wonder if an intermediate branch between incoming and master would be useful for this and other reasons?  i.e. check in to incoming with new features turned on, but when you want to do a release to master, merge first to the intermediate branch, turning off any features that are deemed not yet stable enough.  Then cherry-pick only bug-fixes to that branch until the tests all pass, and finally merge that into master?

**Comment by kripken** [comment]
I revised this proposal, please take a look.

The main change is to admit we only have linux CI coverage currently, and to not let that block us merging to master (or updating the major version number). We should be more careful when doing so, but that amounts to also running tests locally + not having recent big changes or recent known regressions. I think that's the best we can do for now - it will still keep master more stable than incoming, even if it's far from the stability we'd like to have.

**Comment by kripken** [comment]
We apparently had a major MacOS regression in 1.37.36 (#6456), which made me suggest we change the proposed process here a little: major versions should happen after minor versions + nothing interesting for a while.

**Comment by kripken** [comment]
Looks like no concerns here, merging.

**Pull request merged**